src/test/scala/is/hail/io/LoadBgenSuite.scala

```diff
@@ -168,41 +189,49 @@ class LoadBgenSuite extends SparkSuite {
       dataSize += nBitsPerProb
 
       while (dataSize >= 8) {
-        packedInput(byteIndex) = (data & 0xffL).toByte
+        packedInput(nSamples + 10 + byteIndex) = (data & 0xffL).toByte
         data = data >>> 8
         dataSize -= 8
         byteIndex += 1
       }
     }
 
     if (dataSize > 0)
-      packedInput(byteIndex) = (data & 0xffL).toByte
+      packedInput(nSamples + 10 + byteIndex) = (data & 0xffL).toByte
 
     packedInput
   }
 
   object TestProbIterator extends Properties("ImportBGEN") {
-    val probIteratorGen = for (nBitsPerProb <- Gen.choose(1, 32);
-      nProbabilities <- Gen.choose(1, 20);
-      size = ((1L << nBitsPerProb) - 1).toUInt;
-      input <- Gen.partition(nProbabilities, size)
-    ) yield (nBitsPerProb, input)
-
-    property("bgen probability iterator gives correct result") =
-      forAll(probIteratorGen) { case (nBitsPerProb, input) =>
-        val packedInput = bitPack(input, nBitsPerProb)
-        val probIterator = new BgenProbabilityIterator(new ByteArrayReader(packedInput), nBitsPerProb)
-        val result = new Array[UInt](input.length)
-
-        for (i <- input.indices) {
-          if (probIterator.hasNext) {
-            result(i) = probIterator.next()
-          } else {
-            fatal(s"Should have at least ${ input.length } probabilities. Found i=$i")
+    val probIteratorGen = for (
+      nBitsPerProb <- Gen.choose(1, 32);
+      nSamples <- Gen.choose(1, 5);
+      nGenotypes <- Gen.choose(3, 5);
+      nProbabilities = nSamples * (nGenotypes - 1);
+      totalProb = ((1L << nBitsPerProb) - 1).toUInt;
+      sampleProbs <- Gen.buildableOfN[Array, Array[UInt]](nSamples, Gen.partition(nGenotypes - 1, totalProb));
+      input = sampleProbs.flatten
+    ) yield (nBitsPerProb, nSamples, nGenotypes, input)
+
+    property("bgen probability iterator, array give correct result") =
+      forAll(probIteratorGen) { case (nBitsPerProb, nSamples, nGenotypes, input) =>
+        assert(input.length == nSamples * (nGenotypes - 1))
+        val packedInput = bitPack(input, nSamples, nBitsPerProb)
+        val reader = new ByteArrayReader(packedInput)
+        reader.seek(nSamples + 10)
+        val probIter = new BgenProbabilityIterator(reader, nBitsPerProb)
+        val probArr = new BGen12ProbabilityArray(packedInput, nSamples, nGenotypes, nBitsPerProb)
+
+        for (s <- 0 until nSamples)
+          for (i <- 0 until (nGenotypes - 1)) {
+            val expected = input(s * (nGenotypes - 1) + i)
+            assert(probIter.hasNext)
+            assert(probIter.next() == expected)
+            assert(probArr(s, i) == expected)
```

Comment:
Do you still need the `BgenProbabilityIterator` in the tests? I don't see what testing against the iterator is buying you here.